lib/AST/ModuleDependencies.cpp

```diff
@@ -332,24 +291,102 @@ void ModuleDependenciesCache::recordDependencies(
     // Ensure that the existing dependencies objects are at a different path.
     // i.e. do not record duplicate dependencies.
     auto newModulePath = modulePathForVerification(dependencies);
-    bool newEntry = true;
     for (auto &existingDeps : map[moduleName]) {
       if (modulePathForVerification(existingDeps) == newModulePath)
-        newEntry = false;
+        return &existingDeps;
     }
-    if (newEntry)
-      map[moduleName].emplace_back(std::move(dependencies));
+
+    map[moduleName].emplace_back(std::move(dependencies));
+    return map[moduleName].end()-1;
   } else {
     map.insert({moduleName, ModuleDependenciesVector{std::move(dependencies)}});
     AllModules.push_back({moduleName.str(), kind});
+    return &(map[moduleName].front());
   }
 }
 
-void ModuleDependenciesCache::updateDependencies(
+const ModuleDependencies* GlobalModuleDependenciesCache::updateDependencies(
     ModuleDependencyID moduleID, ModuleDependencies dependencies) {
   auto &map = getDependenciesMap(moduleID.second);
   auto known = map.find(moduleID.first);
   assert(known != map.end() && "Not yet added to map");
-  assert(known->second.size() == 1 && "Cannot update dependency with multiple candidates.");
+  assert(known->second.size() == 1 &&
+         "Cannot update dependency with multiple candidates.");
   known->second[0] = std::move(dependencies);
+  return &(known->second[0]);
+}
+
+llvm::StringMap<const ModuleDependencies*> &
+ModuleDependenciesCache::getDependencyReferencesMap(ModuleDependenciesKind kind) {
+  auto it = ModuleDependenciesKindMap.find(kind);
+  assert(it != ModuleDependenciesKindMap.end() && "invalid dependency kind");
+  return it->second;
+}
+
+const llvm::StringMap<const ModuleDependencies*> &
+ModuleDependenciesCache::getDependencyReferencesMap(ModuleDependenciesKind kind) const {
+  auto it = ModuleDependenciesKindMap.find(kind);
+  assert(it != ModuleDependenciesKindMap.end() && "invalid dependency kind");
+  return it->second;
+}
+
+Optional<const ModuleDependencies*> ModuleDependenciesCache::findDependencies(
+    StringRef moduleName, Optional<ModuleDependenciesKind> kind) const {
+  if (!kind) {
+    if (auto swiftTextualDep =
+            findDependencies(moduleName, ModuleDependenciesKind::SwiftTextual))
+      return swiftTextualDep;
+    else if (auto swiftBinaryDep = findDependencies(
+                 moduleName, ModuleDependenciesKind::SwiftBinary))
+      return swiftBinaryDep;
+    else if (auto swiftPlaceholderDep = findDependencies(
+                 moduleName, ModuleDependenciesKind::SwiftPlaceholder))
+      return swiftPlaceholderDep;
+    else
+      return findDependencies(moduleName, ModuleDependenciesKind::Clang);
```

Comment:
Thanks! Additional nitpick: can we define enum cases of `ModuleDependenciesKind::first_Kind` and `ModuleDependenciesKind::last_Kind` so here we could use a for-loop like:
```
for (auto kind = ModuleDependenciesKind::first_Kind; kind != ModuleDependenciesKind::last_Kind; ++ kind) {
  if (auto dep = findDependencies(moduleName, kind)) {
    return dep;
  }
}
return llvm::None;
```
This way so we could avoid updating code here if new dependency kind is added in the future.

Reply:
`ModuleDependenciesCache::ModuleDependenciesCache` could also be implemented without enumerating all cases explicitly.

Reply:
I do not think we should do this. We would lose the benefits of the fact that this is currently a strongly-typed enum which does not implicitly convert to `int`. 
To make this kind of iteration work, we would also need to define `operator++` for `ModuleDependenciesKind`, like so:
```
ModuleDependenciesKind operator++(ModuleDependenciesKind& e)
{
   int i = static_cast<int>(e);
   return static_cast<ModuleDependenciesKind>((i + 1));
}
```
or have casts in the loop condition.
We would then need to ensure that the enum remains contiguous (it doesn't have to), and make sure we avoid overflow. We would also need to be on the lookout to avoid introducing behaviours that start relying on the order of dependency kinds. This all feels quite a bit more fragile than what we have.

To address the concern of such use-sites going un-noticed when adding new dependency kinds, how about we use a `switch` statement instead, without a default, which would get the compiler to force the user to update it when a new dependency kind is added?

Reply:
Ah, we can't use a `switch` here because we don't actually have a value to switch over.

Reply:
I think I can make this work keeping the `enum class` and sufficient error checking in the increment.

Reply:
Done!

Reply:
yeah, `ModuleDependenciesKind` should have `int8_t` as the underlying value to make this operation available.